### ISSUE TITLE
undefined reference to TIFFReadRGBAStrip@LIBTIFF_4.0

### ISSUE THREAD

**Issue by zhangxgu** [facebookarchive/caffe2#1799]
Error came up when building caffe2.  

> [ 76%] Building CXX object caffe2/CMakeFiles/reshape_op_gpu_test.dir/operators/reshape_op_gpu_test.cc.o
> [ 76%] Linking CXX executable ../bin/reshape_op_gpu_test
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFReadRGBAStrip@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFIsTiled@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFWriteScanline@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFGetField@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFScanlineSize@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFReadEncodedTile@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFReadRGBATile@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFClose@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFRGBAImageOK@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFOpen@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFReadEncodedStrip@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFSetField@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFSetWarningHandler@LIBTIFF_4.0'
> /usr/local/lib/libopencv_highgui.so.2.4.13: undefined reference to `TIFFSetErrorHandler@LIBTIFF_4.0'
> collect2: error: ld returned 1 exit status
> caffe2/CMakeFiles/reshape_op_gpu_test.dir/build.make:134: recipe for target 'bin/reshape_op_gpu_test' failed
> make[3]: *** [bin/reshape_op_gpu_test] Error 1
> make[3]: Leaving directory '/home/gzx/caffe2/build'
> CMakeFiles/Makefile2:1284: recipe for target 'caffe2/CMakeFiles/reshape_op_gpu_test.dir/all' failed
> make[2]: *** [caffe2/CMakeFiles/reshape_op_gpu_test.dir/all] Error 2
> make[2]: Leaving directory '/home/gzx/caffe2/build'
> Makefile:138: recipe for target 'all' failed
> make[1]: *** [all] Error 2
> make[1]: Leaving directory '/home/gzx/caffe2/build'
> Makefile:4: recipe for target 'all' failed
> make: *** [all] Error 2

If this is a build issue, please fill out the template below.

### System information

* Operating system: ubuntu 16.04
* Compiler version: gcc 5.4.0
* CMake version:  3.5.1
* CMake arguments:
* Relevant libraries/versions (e.g. CUDA): CUDA 8.0 Cudnn 6.1

### CMake summary output

```
******** Summary ********
There is no summary...
```

**Comment by pjh5** [comment]
This is the output from the `make` step, but before this ran another command `cmake` was run, which contains information about what packages were found on your machine and what libraries it will try to include when building caffe2. It should look like a 2 column list of things like "USE_OPENCV      : ON" and "OpenCV version   : 3.3.1". Can you please find and post this output, along with the commands you used to build caffe2?

Are you using Anaconda? It looks like your opencv was installed by brew. Can you also post the output of:
`brew list`
and
`find / -name libtiff* 2>/dev/null` ?

If there's no libtiff on your system, try installing that with your favorite package manager and then trying to build again.

**Comment by zhangxgu** [comment]
@pjh5
Thank you for advice!
here is the summary:
******** Summary ********
-- General:
--   CMake version         : 3.5.1
--   CMake command         : /usr/bin/cmake
--   Git version           : v0.8.1-978-g6a938bb-dirty
--   System                : Linux
--   C++ compiler          : /usr/bin/c++
--   C++ compiler version  : 5.4.0
--   Protobuf compiler     : /usr/local/bin/protoc
--   Protobuf include path : /usr/local/include
--   Protobuf libraries    : optimized;/usr/local/lib/libprotobuf.so;debug;/usr/local/lib/libprotobuf.so;-pthread
--   CXX flags             :  -std=c++11 -O2 -fPIC -Wno-narrowing -Wno-invalid-partial-specialization
--   Build type            : Release
--   Compile definitions   : 
-- 
--   BUILD_BINARY          : ON
--   BUILD_DOCS            : OFF
--   BUILD_PYTHON          : ON
--     Python version      : 2.7.13
--     Python library      : /home/gzx/anaconda2/lib/libpython2.7.so
--   BUILD_SHARED_LIBS     : ON
--   BUILD_TEST            : ON
--   USE_ATEN              : OFF
--   USE_ASAN              : OFF
--   USE_CUDA              : ON
--     CUDA version        : 8.0
--     CuDNN version       : 6.0.21
--   USE_EIGEN_FOR_BLAS    : 1
--   USE_FFMPEG            : OFF
--   USE_GFLAGS            : ON
--   USE_GLOG              : ON
--   USE_GLOO              : ON
--   USE_LEVELDB           : ON
--     LevelDB version     : 1.18
--     Snappy version      : 1.1.3
--   USE_LITE_PROTO        : OFF
--   USE_LMDB              : ON
--     LMDB version        : 0.9.17
--   USE_METAL             : OFF
--   USE_MKL               : 
--   USE_MOBILE_OPENGL     : OFF
--   USE_MPI               : ON
--   USE_NCCL              : ON
--   USE_NERVANA_GPU       : OFF
--   USE_NNPACK            : ON
--   USE_OBSERVERS         : ON
--   USE_OPENCV            : ON
--     OpenCV version      : 2.4.13.2
--   USE_OPENMP            : OFF
--   USE_PROF              : OFF
--   USE_REDIS             : OFF
--   USE_ROCKSDB           : OFF
--   USE_THREADS           : ON
--   USE_ZMQ               : OFF
-- Configuring done
-- Generating done

and I am not using anaconda,.
The output of brew list:
--The program 'brew' is currently not installed.
The output of find / -name libtiff* 2>/dev/null is :

> -- /usr/local/lib/python2.7/dist-packages/PIL/.libs/libtiff-fa49d64f.so.5.2.6
> /usr/local/MATLAB/R2016b/bin/glnxa64/libtiff.so.5.0.5
> /usr/local/MATLAB/R2016b/bin/glnxa64/libtiff.so.5
> /usr/local/MATLAB/R2016b/mcr/toolbox/matlab/imagesci/libtiffcopyright.txt
> /usr/local/MATLAB/R2016b/toolbox/matlab/imagesci/libtiffcopyright.txt
> /usr/lib/x86_64-linux-gnu/libtiff.so.5.2.4
> /usr/lib/x86_64-linux-gnu/libtiffxx.a
> /usr/lib/x86_64-linux-gnu/libtiff.a
> /usr/lib/x86_64-linux-gnu/libtiff.so.5
> /usr/lib/x86_64-linux-gnu/pkgconfig/libtiff-4.pc
> /usr/lib/x86_64-linux-gnu/libtiffxx.so.5.2.4
> /usr/lib/x86_64-linux-gnu/libtiffxx.so.5
> /usr/lib/x86_64-linux-gnu/libtiff.so
> /usr/lib/x86_64-linux-gnu/libtiffxx.so
> /usr/share/boost-build/src/tools/libtiff.jam
> /usr/share/doc/libtiff5
> /usr/share/doc/libtiffxx5
> /usr/share/doc/libtiff5-dev
> /usr/share/man/man3/libtiff.3tiff.gz
> /usr/share/lintian/overrides/libtiff5
> /usr/share/lintian/overrides/libtiffxx5
> /usr/share/lintian/overrides/libtiff5-dev
> /home/gzx/anaconda2/conda-meta/libtiff-4.0.8-h90200ff_9.json
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiffxx.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiff.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiffxx.so.5.3.0
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiffxx.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiff.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiff.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiffxx.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiffxx.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiff.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0/lib/libtiff.so.5.3.0
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/share/doc/tiff-4.0.6/html/libtiff.html
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/share/doc/tiff-4.0.6/html/man/libtiff.3tiff.html
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiffxx.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiff.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiffxx.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiff.so.5.2.4
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiff.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiff.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiffxx.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiffxx.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiff.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2/lib/libtiffxx.so.5.2.4
> /home/gzx/anaconda2/pkgs/libtiff-4.0.9-h28f6b97_0.tar.bz2
> /home/gzx/anaconda2/pkgs/libtiff-4.0.6-2.tar.bz2
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiffxx.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiff.a
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiffxx.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiff.so
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiff.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiff.so.5.2.6
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiffxx.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiffxx.so.5
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiff.la
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9/lib/libtiffxx.so.5.2.6
> /home/gzx/anaconda2/pkgs/libtiff-4.0.8-h90200ff_9.tar.bz2
> /home/gzx/anaconda2/lib/libtiffxx.a
> /home/gzx/anaconda2/lib/libtiff.a
> /home/gzx/anaconda2/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/lib/libtiffxx.so
> /home/gzx/anaconda2/lib/libtiff.so
> /home/gzx/anaconda2/lib/libtiff.so.5
> /home/gzx/anaconda2/lib/libtiff.so.5.2.6
> /home/gzx/anaconda2/lib/libtiffxx.la
> /home/gzx/anaconda2/lib/libtiffxx.so.5
> /home/gzx/anaconda2/lib/libtiff.la
> /home/gzx/anaconda2/lib/libtiffxx.so.5.2.6
> /home/gzx/anaconda2/envs/denseflow/conda-meta/libtiff-4.0.9-h28f6b97_0.json
> /home/gzx/anaconda2/envs/denseflow/lib/libtiffxx.a
> /home/gzx/anaconda2/envs/denseflow/lib/libtiff.a
> /home/gzx/anaconda2/envs/denseflow/lib/libtiffxx.so.5.3.0
> /home/gzx/anaconda2/envs/denseflow/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/envs/denseflow/lib/libtiffxx.so
> /home/gzx/anaconda2/envs/denseflow/lib/python2.7/site-packages/PIL/.libs/libtiff-fa49d64f.so.5.2.6
> /home/gzx/anaconda2/envs/denseflow/lib/libtiff.so
> /home/gzx/anaconda2/envs/denseflow/lib/libtiff.so.5
> /home/gzx/anaconda2/envs/denseflow/lib/libtiffxx.la
> /home/gzx/anaconda2/envs/denseflow/lib/libtiffxx.so.5
> /home/gzx/anaconda2/envs/denseflow/lib/libtiff.la
> /home/gzx/anaconda2/envs/denseflow/lib/libtiff.so.5.3.0
> /home/gzx/anaconda2/envs/gluon/lib/python2.7/site-packages/PIL/.libs/libtiff-fa49d64f.so.5.2.6
> /home/gzx/anaconda2/envs/py35/conda-meta/libtiff-4.0.6-2.json
> /home/gzx/anaconda2/envs/py35/share/doc/tiff-4.0.6/html/libtiff.html
> /home/gzx/anaconda2/envs/py35/share/doc/tiff-4.0.6/html/man/libtiff.3tiff.html
> /home/gzx/anaconda2/envs/py35/lib/libtiffxx.a
> /home/gzx/anaconda2/envs/py35/lib/libtiff.a
> /home/gzx/anaconda2/envs/py35/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/envs/py35/lib/libtiffxx.so
> /home/gzx/anaconda2/envs/py35/lib/libtiff.so.5.2.4
> /home/gzx/anaconda2/envs/py35/lib/libtiff.so
> /home/gzx/anaconda2/envs/py35/lib/libtiff.so.5
> /home/gzx/anaconda2/envs/py35/lib/libtiffxx.la
> /home/gzx/anaconda2/envs/py35/lib/libtiffxx.so.5
> /home/gzx/anaconda2/envs/py35/lib/libtiff.la
> /home/gzx/anaconda2/envs/py35/lib/libtiffxx.so.5.2.4
> /home/gzx/anaconda2/envs/pytorch/conda-meta/libtiff-4.0.9-h28f6b97_0.json
> /home/gzx/anaconda2/envs/pytorch/lib/libtiffxx.a
> /home/gzx/anaconda2/envs/pytorch/lib/libtiff.a
> /home/gzx/anaconda2/envs/pytorch/lib/libtiffxx.so.5.3.0
> /home/gzx/anaconda2/envs/pytorch/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/envs/pytorch/lib/libtiffxx.so
> /home/gzx/anaconda2/envs/pytorch/lib/libtiff.so
> /home/gzx/anaconda2/envs/pytorch/lib/libtiff.so.5
> /home/gzx/anaconda2/envs/pytorch/lib/libtiffxx.la
> /home/gzx/anaconda2/envs/pytorch/lib/libtiffxx.so.5
> /home/gzx/anaconda2/envs/pytorch/lib/libtiff.la
> /home/gzx/anaconda2/envs/pytorch/lib/libtiff.so.5.3.0
> /home/gzx/anaconda2/envs/tensorflow/conda-meta/libtiff-4.0.6-2.json
> /home/gzx/anaconda2/envs/tensorflow/share/doc/tiff-4.0.6/html/libtiff.html
> /home/gzx/anaconda2/envs/tensorflow/share/doc/tiff-4.0.6/html/man/libtiff.3tiff.html
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiffxx.a
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiff.a
> /home/gzx/anaconda2/envs/tensorflow/lib/pkgconfig/libtiff-4.pc
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiffxx.so
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiff.so.5.2.4
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiff.so
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiff.so.5
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiffxx.la
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiffxx.so.5
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiff.la
> /home/gzx/anaconda2/envs/tensorflow/lib/libtiffxx.so.5.2.4
> /home/gzx/YOLOv2/opencv-3.2.0/3rdparty/libtiff
> /home/gzx/YOLOv2/opencv-3.3.0/3rdparty/libtiff

**Comment by srishti-hike** [comment]
Facing the same issue. Ran the following commands:
git clone --recursive https://github.com/caffe2/caffe2.git && cd caffe2
make && cd build && sudo make install
python -c 'from caffe2.python import core' 2>/dev/null && echo "Success" || echo "Failure"
From: https://caffe2.ai/docs/getting-started.html?platform=mac&configuration=compile

**Comment by pjh5** [comment]
@zhangxgu It seems like you have opencv installed under /usr/local and your opencv relies on tiff, which it assumes will be in the same location as it (also in /usr/local). However all your tiff installations are in anaconda environments, Matlab installations, or in /usr. Any of the following should work
- Installing tiff into /usr/local
- Install using Anaconda (if you want to use opencv, add opencv to both requirements lists in caffe2/conda/meta.yaml):
`
conda create -n caffe2_env
`
`
conda build conda
`
`
conda install caffe2 --use-local
`

**Comment by pjh5** [comment]
@srishti-hike What operating system are you using?

**Comment by zhangxgu** [comment]
@pjh5  I sudo apt-get install tiff in /usr/local, but after that I run make and get this error:
> [  1%] Built target nccl_external
> [  1%] Built target pthreadpool
> [  5%] Built target nnpack
> [  6%] Built target nnpack_reference_layers
> [  6%] Built target gtest
> [  7%] Built target gtest_main
> [  9%] Built target benchmark
> [  9%] Building CXX object third_party/gloo/gloo/CMakeFiles/gloo.dir/common/linux.cc.o
> third_party/gloo/gloo/CMakeFiles/gloo.dir/build.make:134: recipe for target 'third_party/gloo/gloo/CMakeFiles/gloo.dir/common/linux.cc.o' failed
> CMakeFiles/Makefile2:588: recipe for target 'third_party/gloo/gloo/CMakeFiles/gloo.dir/all' failed
> Makefile:138: recipe for target 'all' failed

And 

> conda create -n caffe2_env conda build conda conda install caffe2 --use-local

may have something wrong? Because I get 

> PackageNotFoundError: Packages missing in current channels:
> 
>   - build
> 
> We have searched for the packages in the following channels:
> 
>   - https://repo.continuum.io/pkgs/main/linux-64
>   - https://repo.continuum.io/pkgs/main/noarch
>   - https://repo.continuum.io/pkgs/free/linux-64
>   - https://repo.continuum.io/pkgs/free/noarch
>   - https://repo.continuum.io/pkgs/r/linux-64
>   - https://repo.continuum.io/pkgs/r/noarch
>   - https://repo.continuum.io/pkgs/pro/linux-64
>   - https://repo.continuum.io/pkgs/pro/noarch

**Comment by pjh5** [comment]
@zhangxgu sorry Github messed up my code. There were new lines that it removed. I have fixed it

With regards to the gloo error, there's not really enough output for me to tell what the problem is. The actual error is probably a few lines above what you posted (it's not all together because these build steps run in parallel). When you cloned the Caffe2 repo did you use the --recursive flag? Without further output I'd guess that you don't have gloo on your machine, so it's building it from third_party (which just links to gloo on Github and builds from there), which is the desired behavior. However you need to clone Caffe2 with the --recursive flag, or you can call `git submodule update --init` now to populate all the third_party directories

**Comment by lvpchen** [comment]
use root to build:
 sudo make -j

**Comment by pjh5** [comment]
@zhangxgu has your most recent problem been fixed? 

It seems like the original issue has been solved, so I will close this issue. If you have more problems, please open a new issue on github.
